image_writers_guide/sti_testing.adoc

```diff
@@ -0,0 +1,161 @@
+:data-uri:
+:icons:
+:experimental:
+:toc: macro
+:toc-title:
+
+toc::[]
+
+== Overview
+
+This document covers how as a STI builder image author you can test your
+STI image locally and how you can use the OpenShift build system for automated
+testing and continuous integration.
+
+Read more about the link:sti[STI architecture and requirements] before reading
+any further.
+
+== The Goal
+
+As described in the STI documentation, STI does require 'assemble' and 'run'
+commands to be present in order to successfully execute the STI build.  You can
+also provide the 'save-artifacts' command for archiving the build artifacts and
+'usage' command that prints the usage to console when someone runs the Docker
+image outside the STI.
+
+The goal of testing the STI image is to make sure that all commands described
+above works properly, even if the base Docker image changed or the tooling that
+the commands are using was updated.
+
+== Testing Requirements
+
+The standard location for the 'test' command is 'test/run'. This command is
+invoked by the OpenShift STI image builder and it could be a simple Bash script
+or a static Go binary.
+
+Since the 'test/run' command will perform the STI build, you should have STI
+binary available in your system. See the STI
+https://github.com/openshift/source-to-image/blob/master/README.md#installation[README]
+file and follow the installation instructions.
+
+Since the STI is putting together the application source code and Docker image,
+in order to test it, we would need a sample application source we can use during
+the test to verify the application source gets successfully converted into a
+Docker image. The sample application should be simple, but it should also
+exercise the 'assemble' command.
+
+== Using sti create
+
+The STI tooling comes with powerful generation tools to speed up the process of
+a new STI image creation. The
+https://github.com/openshift/source-to-image/blob/master/docs/cli.md#sti-create:[sti
+create] will produce all necessary STI scripts and testing tools along with the
+Makefile. The generated 'test/run' script is very simple and needs some
+adjusting to be useful, but it should give you a solid ground to start hacking.
+
+One thing you need to keep in mind is that the 'test/run' script produced by
+'sti create' command require the sample application sources inside the
+'test/test-app' directory.
+
+== Testing locally
+
+The easiest way to run the STI image tests locally is to use the generated
+Makefile. In case you did not use the 'sti create' command, you can grab this
+template and replace the IMAGE_NAME with your image name.
+
+Sample Makefile:
+---
+IMAGE_NAME = openshift/ruby-20-centos
+
+build:
+	docker build -t $(IMAGE_NAME) .
+
+.PHONY: test
+test:
+	docker build -t $(IMAGE_NAME)-candidate .
+	IMAGE_NAME=$(IMAGE_NAME)-candidate test/run
+---
+
+== Basic work flow
+
+This section describes the default work flow you should be using for testing the
+STI image builders:
```

Comment:
I should opt-out this step, yeah. The test script assume that there is an image to test built already.